feat(mobile): SAV-895: Update PatientProgramRegistrationDetails screen to show and group conditions by categories

rohan-bes opened the pull request:
### Changes

Ended up creating a whole separate component for the PatientProgramRegistrationConditionsDetailsRow since there was so much custom logic. There's some gnarly sort/splice functions so added some comments explaining what they're doing.

### Deploys

- [x] **Deploy to Tamanu Internal** <!-- #deploy -->

### Remember to...

- ...write or update tests
- ...add UI screenshots and **testing notes** to the Linear issue
- ...add any **manual upgrade steps** to the Linear issue
- ...update the [config reference](https://beyond-essential.slab.com/posts/reference-config-file-0c70ukly), [settings reference](https://beyond-essential.slab.com/posts/reference-settings-0blw1x2q), or any [relevant runbook(s)](https://beyond-essential.slab.com/topics/runbooks-bs04ml6c)
- ...call out additions or changes to **config files** for the deployment team to take note of

<!-- Thank you! -->

Commit:
feat(mobile): SAV-895: Update PatientProgramRegistrationDetails screen to show and group conditions by categories

Commit:
feat(mobile): SAV-895: Updated sort to be alphabetical, category group

NavarroEmilioLuis commented on packages/mobile/App/ui/navigation/screens/patientProgramRegistration/PatientProgramRegistrationDetails.tsx:
```diff
@@ -61,6 +69,142 @@ const DataRow = ({
+
+const PatientProgramRegistrationConditionsDetailsRow = ({ conditions }) => {
+  const { getTranslation } = useTranslation();
+  let conditionComponents;
+  if (!Array.isArray(conditions) || conditions.length < 1) {
+    conditionComponents = <StyledText>—</StyledText>;
+  } else {
```
Hmmm don't love the `let` approach but it works. Also `conditions.length < 1` reads a bit whack? I'd refactor into another function that returns the conditionComponents so you can do early return maybe.

NavarroEmilioLuis commented on packages/mobile/App/ui/navigation/screens/patientProgramRegistration/PatientProgramRegistrationDetails.tsx:
```diff
@@ -61,6 +69,142 @@ const DataRow = ({
+      ),
+    );
+
+    // Add a horizontal line between confirmed or resolved and recordedInErrorSection conditions
```
I thought we hid recorded in error ones? At least that's the behavior in desktop

NavarroEmilioLuis reviewed (commented):
Uhmmm well you had a lot to implement. And obviously this is biased (because I did the other design), but I think you could refactor this into something way simpler and also have the benefit of repeating the logic from desktop. Look at file packages/web/app/views/programRegistry/ConditionSection.jsx for that (using sortBy with lodash, splitting the categories, etc).

LMK what you think

Commit:
Refactor conditions rendering logic to use lodash and match @tamanu/web/ConditionsSection

Commit:
Improved styling on PatientProgramRegistrationDetails screen

Commit:
Made components more DRY

Commit:
Added translations for conditions and condition categories

rohan-bes commented:
> I think you could refactor this into something way simpler and also have the benefit of repeating the logic from desktop. Look at file packages/web/app/views/programRegistry/ConditionSection.jsx for that (using sortBy with lodash, splitting the categories, etc).

Thanks for the tip @NavarroEmilioLuis I took a look and yeah the implementation there is a lot cleaner. I basically copied what's there, though I used lodash to do the groupBy as that seemed even simpler (lemme know if you want me to copy that across in this PR).

Added a couple more commits to make some UI enhancements and clean the code. Also added translation for conditions and categories. :pray:

tcaiger commented on packages/mobile/App/ui/navigation/screens/patientProgramRegistration/PatientProgramRegistrationDetails.tsx:
```diff
@@ -1,63 +1,149 @@
+        value={condition.conditionCategory}
+        enumValues={PROGRAM_REGISTRY_CONDITION_CATEGORY_LABELS}
+      />
+      {`)`}
```
```suggestion
      )
```
Since we are in the template we shouldn't need to use the brackets and template literals, right?

rohan-bes replied:
Oh that's much nicer! I had thought this didn't work since my editor does syntax highlighting on the `(` character :sweat_smile:

tcaiger commented on packages/mobile/App/ui/navigation/screens/patientProgramRegistration/PatientProgramRegistrationDetails.tsx:
```diff
@@ -1,63 +1,149 @@
+        value={condition.programRegistryCondition.id}
+        category="programRegistryCondition"
+      />
+      {` (`}
```
```suggestion
{' '}
    (
```
This one is subjective but I find this easier to understand

tcaiger commented on packages/mobile/App/ui/navigation/screens/patientProgramRegistration/PatientProgramRegistrationDetails.tsx:
```diff
@@ -1,63 +1,149 @@
+  );
+  const needsDivider = groupedConditions.closed && groupedConditions.open;
+
+  const TranslatedCondition = ({ condition }) => (
```
Any reason this can't be defined outside of the main component above so that it only gets created once?

rohan-bes replied:
Ah good point, will move out

tcaiger reviewed (approved):
@rohan-bes Looks good to me. I'm happy for you to merge when you are ready

Commit:
Code cleanups

Commit:
Merge branch 'feat/sav-904-program-registry-related-conditions-category-ui-mobile' into feat/sav-895-program-registration-condition-categories

Commit:
Merge branch 'feat/sav-904-program-registry-related-conditions-category-ui-mobile' into feat/sav-895-program-registration-condition-categories

Commit:
Merge branch 'feat/sav-904-program-registry-related-conditions-category-ui-mobile' into feat/sav-895-program-registration-condition-categories

Commit:
Merge branch 'feat/sav-904-program-registry-related-conditions-category-ui-mobile' into feat/sav-895-program-registration-condition-categories

Commit:
fixed condition category translation on mobile

Commit:
SAV-895: Fix test issues
- Flag required fields
- Add Back button warning on Category screen
- Fix labels
